lib/segment/src/index/field_index/full_text_index/inverted_index.rs

```diff
@@ -154,21 +144,45 @@ impl InvertedIndex {
             // Empty request -> no matches
             return Box::new(vec![].into_iter());
         }
-        intersect_postings_iterator(postings)
+
+        match self {
+            InvertedIndex::Mutable(_index) => {
+                // in case of mutable index, deleted documents are removed from the postings
+                intersect_postings_iterator(postings, |_| true)
+            }
+            InvertedIndex::Immutable(index) => {
+                // in case of immutable index, deleted documents are still in the postings
+                let filter = move |idx| {
+                    matches!(
+                        index.point_documents_tokens.get(idx as usize),
+                        Some(Some(_))
+                    )
+                };
+                intersect_postings_iterator(postings, filter)
+            }
+        }
```

Comment:
Can we remove all of this logic into the (im)mutable index themselves, so we don't have to explicitly filter here?

It looks like the immutable structure already has al information for this, the postings and document tokens. Or am I missing an important detail here?

Reply:
You are right, common logic between mutable and immutable states is only `None` conditions. Moved to separate functions, it looks better. Fixed